numpy/_core/numeric.py

```diff
@@ -2341,8 +2341,7 @@ def within_tol(x, y, atol, rtol):
         with errstate(invalid='ignore'), _no_nep50_warning():
             return less_equal(abs(x-y), atol + rtol * abs(y))
 
-    x = asanyarray(a)
-    y = asanyarray(b)
+    x, y, atol, rtol = np.broadcast_arrays(a, b, atol, rtol, subok=True)
```

Comment:
Hmmmpf, this is one of the cases similar to gh-24957.  So, `atol` and `rtol` stop being Python scalars (with NEP 50) or 0-d objects (without NEP 50) and that changes promotion. 

We need a solution for that, right now, we could skip it add logic for them being `type(atol) in (int, float, complex)`.  The question remains a bit whether we want a way to say `mark_was_weak_scalar(array)`, to turn on weak promotion for an *array*, but... I don't really want such a mark to survive for long and it is unclear what operations should retain it.

Reply:
I would think that if NumPy is hitting this, then other libraries and users will hit it as well. Maybe we should rethink the transition and how we provide (or not) backward incompatibility shims.

Reply:
I guess the casting could be done earlier - something like 
```
dtype=np.result_type(a, b, atol, rtol)
atol = np.asanyarray(atol, dtype)
rtol = np.asanyarray(rtol, dtype)
```
and then broadcast?

If broadcast is a common culprit, we might add a new keyword argument that does this implicitly?

Reply:
> If broadcast is a common culprit, we might add a new keyword argument that does this implicitly?

Yes, please! This is something Sebastian and I have discussed before.

Reply:
Yes, that would be the suggestion: to allow the dtype of arrays created from python scalars to follow NEP 50 promotion rules.

Reply:
There are different approaches:
* just make the current stuff a bit more convenient.
* Try to tag on the info to an array.
* Just deal with it here, the `isinstance` check and special code paths are not that terrible (and also useful for micro optimization probably).

To me no solution is obvious, unfortunately.  Tagging the info on to the array or dtype, means that you have a very tricky situation of how to drag the info around and mainly about when to preserve and when to drop it.
In this case for example, you are calling `arr[boolean_idx]` creating a copy, should that drop to prevent accidentally returning such an array?  Assignments at least could be prevent by making it readonly, but...  Also the special path to make `arr > -1` work for such an array is another tricky thing.
I can somewhat think of ways to do this, but I am not sure they will be very pleasant.

Moving resolution to earlier (maybe with a helper) can help.  But in some cases might also not be quite what we want because it makes sense to ask `isclose(2 meters, 1.9 meters, atol=0.1, rtol=0.1 meters)`.  But it is less clear that the: `dtype = result_type(meters, float)` is reasonable (for the `atol`)?